site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,300 @@
+Receiving the redirect tells us that the user has been logged out of FusionAuth and we are safe to clear the user's session in Rails.
+
+### Polishing up our Rails session
+Now that we have the functionality to authenticate users as well as log them out of FusionAuth, we will want to complete the same cycle on our Rails app.
+
+There are numerous ways to handle user sessions in Rails, but for this example, I decided to create helper methods accessible to all sub-classes of ApplicationController.
+
+`current_user` looks for a user jwt on the session. If it exists, it decodes the JWT and retrieves user's email address.
+If it does not exist, this means the user has logged out or their session has expired.
+
+```ruby
+# app/controllers/application_controller.rb
+
+class ApplicationController < ActionController::Base
+  helper_method :current_user
+  helper_method :logged_in?
+
+  def current_user
+    if session[:user_jwt]
+      token = session[:user_jwt]
+      decoded = TokenService.decode(token["value"])
+      decoded.first["email"]
+    end
+
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+end
+```
+
+```ruby
+# app/services/token_service.rb
+
+class TokenService
+  def self.decode(token)
+    JWT.decode(
+        token,
+        nil,
+        false,
+        {algorithm: 'HS256'})
+  end
+end
+```
+
+## Putting it all together
+
+Kick the tires and light the fires
+```
+rails s
+```
+
+Navigating to `http://localhost:3000`
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/home.png" alt="The example Home screen for the Rails application." class="img-fluid" figure=false %}
+
+Login to the Application
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/fa-login.png" alt="The FusionAuth login page." class="img-fluid" figure=false %}
+
+We made it! We are logged in and our application knows who we are.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/logged-in.png" alt="The example Home screen when logged-in to the Rails application." class="img-fluid" figure=false %}
+
+Logout.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/logging-out.png" alt="The FusionAuth logging out page." class="img-fluid" figure=false %}
+
+## Next Steps
+Now that we have the OAuth flow working, our foundation is set to expand as needed. Here are a few ideas to tackle next:
+* Add and assign user roles such that functionality in our Rails application can be shown or hidden accordingly.
+* Customize the FusionAuth login page with a look and feel of our application including a logo.
+* Add a "Login with Google" or "Login with Facebook" social login.
+
+## What did we learn?
+Using the Authorization Code grant in Rails lets you use any OAuth compatible identity provider to secure your application. The example code can be found on Github [here](https://github.com/FusionAuth/fusionauth-rails-app). 
```

Comment:
Need to make sure to rename the github repo to the proper format, and update the URL here.